Cache docMatchTree results for Meta conditions

binarymason opened the pull request:
At GitLab, we construct large and complicated queries based on user authorizations that include multiple `OR` conditions. Many of these nested individual `OR` conditions are identical across the query.

Currently, when building a new query match tree, we recompute these identical conditions each time they appear, leading to unnecessary CPU usage. This PR introduces caching for these conditions, computing each unique condition only once and reusing the result throughout the match tree construction process.

This optimization reduces redundant computation and prevents CPU usage spikes when constructing complex queries.

Commit:
Cache docMatchTree results for RepoIds and Meta

When constructing a complicated match tree, this cache
allows zoekt to not have to recompute conditions that it has
already seen before.

binarymason commented:
👋 Hey there @keegancsmith. May I ask for a review here? Thanks!

keegancsmith reviewed (changes requested):
Some overall comments to understand more and to fix bugs

- docMatchTree is stateful. You can't use it concurrently. You need to copy it on cache access and reset the fields marked as mutable in it. This should work then and still be fast.
- this isn't a cache that can have evictions. I would consider using something which will evict if certain memory or length values are hit.
- if users control the input at all here, you can probably denial of service relatively easily.
- If in your case it is controlled and you want to keep it simple, make this feature optional and default to off. envvar seems fine here, we do that a few times in sourcegraph over time.
- for example sourcegraph uses the repoid field, and I wouldn't accept this cache since the cache likely will have low hit rate + use up memory. Did you find in testing this improved it for repoids? I'm kinda surprised s.Repos.Contains has that much overhead. TBH looking at the code now we can remove the reposWant and instead do a different predicate: `return s.Repos.Contains(d.repos[docID])`. I think this pattern here predates us having a generic predicate.
- don't use sha unless you need the cryptographic stuff. Use something fast and unlikely to have collisions. eg `github.com/cespare/xxhash` is already a transitive dependency which would be appropriate

Commit:
Use xxhash

Commit:
Use LRU cache

binarymason commented on index/matchtree.go:
```diff
@@ -1435,3 +1457,18 @@ func isRegexpAll(r *syntax.Regexp) bool {
+		b.WriteString(fmt.Sprint(r.ID))
+		b.WriteByte(',')
+	}
+	sum := sha256.Sum256([]byte(b.String()))
```
This makes sense. Updated. Thanks!

Commit:
Specify cache length with ZOEKT_MATCHTREE_CACHE

binarymason commented:
@keegancsmith thanks for the review. I'll do my best to respond inline. Back to you. 🏓 

> * docMatchTree is stateful. You can't use it concurrently. You need to copy it on cache access and reset the fields marked as mutable in it. This should work then and still be fast.

Check 👍 

> * this isn't a cache that can have evictions. I would consider using something which will evict if certain memory or length values are hit.

Great call out! I implemented a simple LRU cache for this.

> * if users control the input at all here, you can probably denial of service relatively easily.

Users cannot control the input at all here.

> * If in your case it is controlled and you want to keep it simple, make this feature optional and default to off. envvar seems fine here, we do that a few times in sourcegraph over time.

Great idea. I gated this feature behind the environment variable `ZOEKT_MATCHTREE_CACHE`. That will allow users to opt-in and specify how many cache items to keep. 

> * for example sourcegraph uses the repoid field, and I wouldn't accept this cache since the cache likely will have low hit rate + use up memory. Did you find in testing this improved it for repoids? I'm kinda surprised s.Repos.Contains has that much overhead. TBH looking at the code now we can remove the reposWant and instead do a different predicate: `return s.Repos.Contains(d.repos[docID])`. I think this pattern here predates us having a generic predicate.

We haven't tested `repoids` in isolation. Our queries are mostly `query_string`, `repoids` and several `meta` conditions. We haven't noticed any major performance problems when testing any of these conditions in isolation. Our gotcha is complicated `or` and `and` children which causes each of these to be re-evaluated in the match tree. 

> * don't use sha unless you need the cryptographic stuff. Use something fast and unlikely to have collisions. eg `github.com/cespare/xxhash` is already a transitive dependency which would be appropriate

Done. ✅

keegancsmith commented on index/matchtree.go:
```diff
@@ -971,6 +974,19 @@ func (d *indexData) newMatchTree(q query.Q, opt matchTreeOpt) (matchTree, error)
+				cacheSize = 0
+			}
+		}
+		d.docMatchTreeCache = newDocMatchTreeCache(cacheSize)
```
minor: this will race on first creation. in practice it is probably fine, but maybe should just always create it with size zero?

keegancsmith commented on index/lrucache.go:
```diff
@@ -0,0 +1,69 @@
+	"container/list"
+)
+
+type LRUCache[K comparable, V any] struct {
```
minor, you only have one implementation of this the generics don't really make it any clearer. maybe just directly call this docMatchTreeCache then it can provide exactly the API you need. IE only export Get and Add, and make them work in the way we expect.

Can even move the cache size envvar into here just to make it all contained in this file.

keegancsmith commented on index/matchtree.go:
```diff
@@ -1201,19 +1225,28 @@ func (d *indexData) newMatchTree(q query.Q, opt matchTreeOpt) (matchTree, error)
+		cacheKeyField := "RepoIDs"
+		if cached, ok := d.docMatchTreeCache.get(cacheKeyField, checksum); ok {
+			return cached, nil
+		}
```
if repoids didn't show up in your profiling/testing, can we avoid caching in it? IE avoid caching unless its shown to be effective.

If it did show up, do you mind experimenting with what I proposed around updating the predicate we use to be smarter (ie avoid the intermediate reposWant slice)

binarymason replied:
> if repoids didn't show up in your profiling/testing, can we avoid caching in it? IE avoid caching unless its shown to be effective.

Sure, that makes sense. I went ahead and removed `repoids` from this PR. Happy to circle back with another iteration of this that includes `repoids` if we can report a performance improvement. 🤝

binarymason committed:
Avoid intermediate allocations

Co-authored-by: Keegan Carruthers-Smith <keegan.csmith@gmail.com>

Commit:
Remove cache for repoids

Commit:
Refactor with reviewer feedback

Commit:
Avoid race for docMatchTreeCache

Commit:
Add more documentation

binarymason commented:
Thanks for the great feedback, @keegancsmith.  It's great to collaborate with you! 🤝 

 I believe I addressed all of your comments/suggestions and this is ready for another look. 🏓

Commit:
Improve comment wording

keegancsmith committed:
make cache thread safe

I missed this at first so I followed up with a commit for ya. At first I
just added RWMutex in, but it was a bit complicated due to Get's needing
to do writes. So I changed the eviction policy from LRU to random. Given
you will be tuning your cache to be big enough for your use case I
imagine this will be alright?

https://ampcode.com/threads/T-10ce9d15-0e69-4903-b99e-c38459fa9219

keegancsmith reviewed (approved):
Hey I missed that docMatchTreeCache was not concurrency safe, so I pushed https://github.com/sourcegraph/zoekt/pull/982/commits/005e3c5da503f06949702da9f77d56c0be7cf6d9 since I am the one who encouraged the LRU in the first place :)

Please take a look at that, if you are happy I'll land this PR.

Thanks for contributing!

binarymason commented:
> Hey I missed that docMatchTreeCache was not concurrency safe, so I pushed https://github.com/sourcegraph/zoekt/commit/005e3c5da503f06949702da9f77d56c0be7cf6d9 since I am the one who encouraged the LRU in the first place :)
> 
> Please take a look at that, if you are happy I'll land this PR.

@keegancsmith your changes look great! Thanks for collaborating with me on this one. 🤝
